pkg/schemacompat/schemacompat.go

```diff
@@ -0,0 +1,374 @@
+	apiextensionsv1.Convert_v1_JSONSchemaProps_To_apiextensions_JSONSchemaProps(existing, &existingInternal, nil)
+	apiextensionsv1.Convert_v1_JSONSchemaProps_To_apiextensions_JSONSchemaProps(new, &newInternal, nil)
+	newStrucural, err := schema.NewStructural(&newInternal)
+	if err != nil {
+		return nil, utilerrors.NewAggregate([]error{err})
+	}
+
+	existingStructural, err := schema.NewStructural(&existingInternal)
+	if err != nil {
+		return nil, utilerrors.NewAggregate([]error{err})
+	}
+
+	lcdStructural := existingStructural.DeepCopy()
+	errs := lcdForStructural(fldPath, existingStructural, newStrucural, lcdStructural, narrowExisting)
+	if len(errs) > 0 {
+		return nil, errs.ToAggregate()
+	}
+	serialized, err := json.Marshal(lcdStructural.ToGoOpenAPI())
+	if err != nil {
+		return nil, utilerrors.NewAggregate([]error{err})
+	}
+	var jsonSchemaProps apiextensionsv1.JSONSchemaProps
+	json.Unmarshal(serialized, &jsonSchemaProps)
+	if err != nil {
+		return nil, utilerrors.NewAggregate([]error{err})
+	}
+	return &jsonSchemaProps, nil
+}
+
+func checkTypesAreTheSame(fldPath *field.Path, existing, new *schema.Structural) (errorList field.ErrorList) {
+	if new.Type != existing.Type {
+		return field.ErrorList{field.Invalid(fldPath.Child("type"), new.Type, "The type of the should not be changed")}
+	}
+	return nil
+}
+
+func checkUnsupportedValidation(fldPath *field.Path, existing, new interface{}, validationName, typeName string) (errorList field.ErrorList) {
+	if !reflect.ValueOf(existing).IsZero() || !reflect.ValueOf(new).IsZero() {
+		return field.ErrorList{field.Forbidden(fldPath, fmt.Sprintf("The '%s' JSON Schema construct is not supported by the Schema negotiation for type '%s'", validationName, typeName))}
+	}
+	return nil
+}
+
+func checkUnsupportedValidationForNumerics(fldPath *field.Path, existing, new *schema.ValueValidation, typeName string) (errorList field.ErrorList) {
+	errorList = append(errorList, checkUnsupportedValidation(fldPath, existing.Not, new.Not, "not", typeName)...)
+	errorList = append(errorList, checkUnsupportedValidation(fldPath, existing.AllOf, new.AllOf, "allOf", typeName)...)
+	errorList = append(errorList, checkUnsupportedValidation(fldPath, existing.AllOf, new.AllOf, "anytOf", typeName)...)
+	errorList = append(errorList, checkUnsupportedValidation(fldPath, existing.AllOf, new.AllOf, "oneOf", typeName)...)
+	errorList = append(errorList, checkUnsupportedValidation(fldPath, existing.Enum, new.Enum, "enum", typeName)...)
+	if !(new.Maximum == existing.Maximum && new.Minimum == existing.Minimum && new.ExclusiveMaximum == existing.ExclusiveMaximum && new.ExclusiveMinimum == existing.ExclusiveMinimum) {
+		errorList = append(errorList, checkUnsupportedValidation(fldPath, existing.Maximum, new.Maximum, "maximum", typeName)...)
+		errorList = append(errorList, checkUnsupportedValidation(fldPath, existing.Minimum, new.Minimum, "minimum", typeName)...)
+	}
+	if new.MultipleOf != existing.MultipleOf {
```

Comment:
Is this `if` necessary? Won't `checkUnsupportedValidation` also perform this check?

Reply:
It is necessary, since `checkUnsupportedValidation` is a generic function that returns an error message if at least one the 2 values in non-zero.
In addition to that basic check, here we're comparing the float values... Well, we should, but it seems I have a bug here since I'm comparing pointers...
I'll fix that.